ignite/pkg/availableport/availableport.go

```diff
@@ -7,17 +7,74 @@ import (
 	"time"
 )
 
+type OptionalParameters struct {
+	WithRandomizer *rand.Rand
+	WithMinPort    int
+	WithMaxPort    int
+}
+
 // Find finds n number of unused ports.
 // it is not guaranteed that these ports will not be allocated to
 // another program in the time of calling Find().
-func Find(n int) (ports []int, err error) {
-	min := 44000
-	max := 55000
+func Find(n int, moreParameters ...OptionalParameters) (ports []int, err error) {
+	// Defining them before so we can set a value depending on the OptionalParameters
+	var min int
+	var max int
+	var r *rand.Rand
 
+	if len(moreParameters) != 0 {
+		extra := moreParameters[0]
+		if extra.WithMinPort != 0 {
+			if extra.WithMinPort > -1 {
+				min = extra.WithMinPort
+			} else {
+				// This is not required since the port would become 0
+				// but the user could not notice that sent a negative port
+				return nil, fmt.Errorf("ports can't be negative (negative min port given)")
```

Comment:
please, try to avoid to much nested conditions

Reply:
After the edition of the formatting you mentioned the nested conditions are now less. Should I reduce them more than how it's now?

Reply:
```go
package availableport

import (
	"fmt"
	"math/rand"
	"net"
	"time"
)

type availablePortOptions struct {
	randomizer *rand.Rand
	minPort    uint
	maxPort    uint
}

type Options func(o *availablePortOptions)

func WithRandomizer(r *rand.Rand) Options {
	return func(o *availablePortOptions) {
		o.randomizer = r
	}
}

func WithMaxPort(maxPort uint) Options {
	return func(o *availablePortOptions) {
		o.maxPort = maxPort
	}
}

func WithMinPort(minPort uint) Options {
	return func(o *availablePortOptions) {
		o.minPort = minPort
	}
}

// Find finds n number of unused ports.
// it is not guaranteed that these ports will not be allocated to
// another program in the time of calling Find().
func Find(n uint, options ...Options) (ports []uint, err error) {
	// Apply the options
	opts := availablePortOptions{
		minPort:    44000,
		maxPort:    55000,
		randomizer: rand.New(rand.NewSource(time.Now().UnixNano())),
	}
	for _, apply := range options {
		apply(&opts)
	}
	// If the number of ports required is bigger than the range, this stops it
	if opts.maxPort < opts.minPort {
		return nil, fmt.Errorf("invalid ports range: max < min (%d < %d)", opts.maxPort, opts.minPort)
	}

	// If the number of ports required is bigger than the range, this stops it
	if n > (opts.maxPort - opts.minPort) {
		return nil, fmt.Errorf("invalid amount of ports requested: limit is %d", opts.maxPort-opts.minPort)
	}

	// Marker to point if a port is already added in the list
	registered := make(map[uint]bool)
	for i := uint(0); i < n; i++ {
		for {
			// Greater or equal to min and lower than max
			totalPorts := opts.maxPort - opts.minPort + 1
			randomPort := opts.randomizer.Intn(int(totalPorts))
			port := uint(randomPort) + opts.minPort

			conn, err := net.Dial("tcp", fmt.Sprintf(":%d", port))
			// if there is an error, this might mean that no one is listening from this port
			// which is what we need.
			if err == nil {
				conn.Close()
				continue
			}
			defer conn.Close()

			// if the port is already registered we skip it to the next one
			// otherwise it's added to the ports list and pointed in our map
			if registered[port] {
				continue
			}
			ports = append(ports, port)
			registered[port] = true
			break
		}
	}
	return ports, nil
}
```

you are doing the parameters verifications more than one time, the nested conditions are not necessary